src/entity.rs

```diff
@@ -125,7 +125,7 @@ where
 
     /// Every row can create few combinations (cartesian products) which are hashed and provided for sparse matrix creation.
     /// `row` - array of strings such as: ("userId1", "productId1 productId2", "brandId1").
-    pub fn process_row(&mut self, row: Vec<&str>) {
+    pub fn process_row(&mut self, row: Vec<SmallVec<[String; SMALL_VECTOR_SIZE]>>) {
```

Comment:
Will be nice to don't use Strings - creating new one every time degrade performance. You can compare performance between impl.